DeploymentAgentReconfigurationScripts/TFS2018 Update2/HardwareMoveRemoveOldAgent.ps1

```diff
@@ -0,0 +1,178 @@
+$collectionName = "NoAbleToReadAgentFile"
+$projectName = "NoAbleToReadAgentFile"
+$deploymentGroupId = "NoAbleToReadAgentFile"
+$deploymentGroupName = "NoAbleToReadAgentFile"
+$agentName = "NoAbleToReadAgentFile"
+$tags = "NoAbleToReadAgentFile"
+$deploymentMachineId = "NoAbleToReadAgentFile"
+
+foreach($line in Get-Content .\.agent) {
+    $token = $line.split('"');
+
+    if ($token[1] -eq 'serverUrl'){
+        $sourceTFSUrl = $token[3];
+        if ($sourceTFSUrl.StartsWith($targetTFSUrl)){
+            Write-Verbose -Verbose "Agent is already configured to the TFS $targetTFSUrl";
+            return 0;
+        }
+    }
+
+    if ($token[1] -eq 'collectionName'){
+        $collectionName = $token[3];
+    }
+
+    if ($token[1] -eq 'projectId'){
+        $projectName = $token[3];
+    }
+
+    if ($token[1] -eq 'deploymentGroupId'){
+        $deploymentGroupId = $token[2].Substring(2, $token[2].Length - 3);
+    }
+
+    if ($token[1] -eq 'agentName'){
+        $agentName = $token[3];
+    }
+}
+
+# Get the Deployment group name
+$encodedPat = [System.Convert]::ToBase64String([System.Text.Encoding]::UTF8.GetBytes(":$patToken"))
+$getDeploymentGroupUrl = $targetTFSUrl + '/' + $collectionName + '/' + $projectName + '/_apis/distributedtask/deploymentgroups/' + $deploymentGroupId + '?api-version=5.0-preview.1'
+ 
+$dg = Invoke-RestMethod -Uri $getDeploymentGroupUrl -Method Get -Headers @{Authorization = "Basic $encodedPat"} -ContentType "application/json"
+$deploymentGroupName = $dg.name
+
+
+# Get the Machine Tags
+$getDeploymentMachineUrl = $targetTFSUrl + '/' + $collectionName + '/' + $projectName + '/_apis/distributedtask/deploymentgroups/' + $deploymentGroupId + '/targets?name=' + $agentName + '&api-version=5.0-preview.1'
+
+$dm = Invoke-RestMethod -Uri $getDeploymentMachineUrl -Method Get -Headers @{Authorization = "Basic $encodedPat"} -ContentType "application/json"
+$tags = $dm.value[0].tags -join ",";
+$deploymentMachineId = $dm.value[0].id;
+
+if ($dm.value[0].agent.status -ne 'offline'){
+    Write-Verbose -Verbose "Agent is already configured to the TFS $targetTFSUrl"
+    return 0;
+}
+
+
+# Create and go to new agent folder
+If(-NOT (Test-Path $env:SystemDrive\'vstsagent'))
+{
+    mkdir $env:SystemDrive\'vstsagent'
+}
+
+$newAgentFolfer = "";
+
+cd $env:SystemDrive\'vstsagent';
+for ($i=1; $i -lt 100; $i++){
+    $destFolder="A"+$i.ToString();
+    if (-NOT (Test-Path ($destFolder))){
+        $newAgentFolfer = $destFolder;
+        break;
+    }
+};
+$newAgentPath = $env:SystemDrive + '\vstsagent\' + $newAgentFolfer;
+
+if ($action -ne "apply"){
+    Write-Verbose -Verbose "If action is set to apply, this script will delete the existing unusable, offline agent $agentName from the deployment group $deploymentGroupName in the TFS $targetTFSUrl and re-configure a new agent in $newAgentPath path with same name and properties including tags $tags. All the old deployment histroty will be deleted.";
```

Comment:
deployment history associated with old agent id will be removed along with agent removal.